sycl/doc/extensions/Uniform/Uniform.asciidoc

```diff
@@ -0,0 +1,190 @@
+
+:blank: pass:[ +]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+== Introduction
+IMPORTANT: This specification is a draft.
+
+NOTE: Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by permission by Khronos.
+
+NOTE: This document is better viewed when rendered as html with asciidoctor.  GitHub does not render image icons.
+
+This extension introduces a mechanism to assert that the value of a variable is
+the same for all work-items in a sub-group.
+
+== Notice
+
+Copyright (c) 2021 Intel Corporation.  All rights reserved.
+
+== Status
+
+Working Draft
+
+This is a preview extension specification, intended to provide early access to a feature for review and community feedback. When the feature matures, this specification may be released as a formal extension.
+
+Because the interfaces defined by this specification are not final and are subject to change they are not intended to be used by shipping software products.
+
+== Version
+
+Built On: {docdate} +
+Revision: 1
+
+== Contacts
+
+John Pennycook, Intel (john 'dot' pennycook 'at' intel 'dot' com) +
+Roland Schulz, Intel (roland 'dot' schulz 'at' intel 'dot' com) +
+Jason Sewall, Intel (jason 'dot' sewall 'at' intel 'dot' com) +
+
+== Dependencies
+
+This extension is written against the SYCL 2020 specification, Revision 2.
+
+== Feature Test Macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification section 6.3.3 "Feature test macros".  Therefore, an
+implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_UNIFORM` to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's APIs the implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version.  Base features are supported.
+|===
+
+== Overview
+
+Some implementations of SYCL map work-items in the same sub-group to SIMD
+hardware. The performance of such implementations can be improved by knowledge
+of how the values held by each work-item relate to one another. In the general
+case, an implementation must assume that the values on each work-item may
+differ (i.e. that the values are _varying_); if an implementation can prove
+that the values on each work-item are the same (i.e. that they are _uniform_),
+it may be able to optimize more aggressively or avoid redundant computation.
+
+These concepts are already exposed in other SPMD programming models:
+
+- OpenMP's `declare simd` directive has a `uniform` clause
+- ISPC's `uniform` qualifier makes uniformity part of the type system
```

Comment:
You can't simply say "this is uniform", something is needed to ensure that in compile-time, or at least at run-time. E.g. "uniform" in OpenMP is ensured by that it will be never used in context when the function parameter is non-uniform.  In ISPC it's a base property of the object, it's created as uniform. 
Just creating uniform from anything can lead to hard run-time errors when "uniform" is used incorrectly. Do we want users to create such a checks at run time? How is that possible? Or are we going to provide some automatic functions?

Reply:
OpenMP and ISPC are only mentioned here to highlight that the concept of uniformity is important; we're not trying to recreate their semantics in SYCL.  They're very different languages, and the solutions they've adopted don't work for SYCL.  OpenMP's solution is tied to SIMD execution, and the concept of uniformity only exists when the compiler is performing loop or function vectorization.  The concept of uniformity in ISPC does more than state that the values are the same, and requests that  a single scalar mutable variable be shared between all SIMD lanes executing a vectorized program.

The concept of uniformity that we are proposing here is closer to the idea of uniformity in SPIR-V (see https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_uniformity_a_uniformity).  A variable is uniform if all work-items have the same value -- a uniform variable is not required to be stored in a scalar register, although implementations are free to employ this optimization if they wish.

I'm not saying that a user cannot shoot themselves in the foot by asserting that a value is uniform when it isn't, but this sort of thing exists in many places already.  If a user asserts that a pointer is aligned (using https://en.cppreference.com/w/cpp/memory/assume_aligned) and it isn't true at run-time, they will also get undefined behavior.

Reply:
A user will not shoot himself if we don't give them a gun that can allow such shooting. We at least should give a ways to check that it's really uniform.

Reply:
There is already a way to check that a variable has the same value of `x` for all work-items in some group `g`:
```c++
sycl::all_of_group(g, (x == sycl::group_broadcast(g, x)))
```

Reply:
It seems easy anyway to have `uniform` returning a wrapping type which is interpreted by `invoke_simd` to verify it is actually uniform at runtime with a debug mode of the runtime.
If at the end it doubles the performance on some hardware while having some checking mode possible, I can accept this SIMD foot gun inside SYCL. :-)
It is not worse than a C++ user using the wrong parallel policy in C++20 algorithms.

Reply:
> It seems easy anyway to have `uniform` returning a wrapping type which is interpreted by `invoke_simd` to verify it is actually uniform at runtime with a debug mode of the runtime.

Completely agree.  Nothing is stopping implementations from providing this behavior if they want, but since the performance overhead of checking every `uniform` declaration may be quite high, we shouldn't bake it in to the specification as a required default behavior.